### PR TITLE
Closes #448 | Add/Update Dataloader alorese

### LINKED ISSUE #448: Create dataset loader for Alorese Collection
Dataloader name: `alorese/alorese.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?alorese

|     Dataset| alorese |
|-------------|---|
| Description | Alorese Collection or Alorese Corpus is a collection of language data in a couple of Alorese variation (Alor and Pantar Alorese). The collection is available in video, audio, and text formats with genres ranging from Experiment or task, Stimuli, Discourse, and Written materials. |
| Subsets | - |
| Languages | aol, ind |
| Tasks | Language Modeling, Automatic Speech Recognition, Machine Translation |
| License | Unknown (unknown) |
| Homepage | https://hdl.handle.net/1839/e10d7de5-0a6d-4926-967b-0a8cc6d21fb1 |
| HF URL | - |
| Paper URL | https://scholarlypublications.universiteitleiden.nl/handle/1887/70891 |

### PR DESCRIPTION
**Pull request by patrickamadeus**
Closes #448 

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/{my_dataset}/{my_dataset}.py` (please use only lowercase and underscore for dataset folder naming, as mentioned in dataset issue) and its `__init__.py` within `{my_dataset}` folder.
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_LOCAL`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py` or `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py --subset_id {subset_name_without_source_or_seacrowd_suffix}`.
- [ ] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### T2T
image

### SPTEXT
image

### SPTEXT_TRANS
image

### CODE REVIEW

File: seacrowd/sea_datasets/alorese/alorese.py
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/alorese/alorese.py:124]: I suggest an `else` statement here that raises `ValueError` if the schema isn't one of the presets we have

**Review by ljvmiranda921 — COMMENTED** [review]
I'm ok with the implementation, just have a few comments. Can you please run `make check_file=...` just so that the formatting is consistent everywhere?

File: seacrowd/sea_datasets/alorese/alorese.py
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/alorese/alorese.py:28]: Not sure what's the best way to import this. My gut feel is that we should use absolute imports here because we're importing the `seacrowd` modules the same way. I don't have strong opinions though

File: seacrowd/sea_datasets/alorese/alorese_url.py
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/alorese/alorese_url.py:1]: Just curious: how were you able to generate this dictionary? Do you think it's possible to automate this process instead instead of keeping this file?

Comment (patrickamadeus) [reply seacrowd/sea_datasets/alorese/alorese_url.py:1]: Hi @ljvmiranda921  ! Based on the [archive](https://archive.mpi.nl/tla/islandora/object/tla:1839_1cd5563b_cfb4_4b0b_bc05_a18babddc86e?asOfDateTime=2019-03-01T10:32:30.790Z), here are the steps:

1. I scraped through the pagination to grab all of the blob url that leads to the [detail page](https://archive.mpi.nl/tla/islandora/object/tla%3A1839_77a7f739_3f6f_4ce4_b6ef_916e3825fa6a)
2. From each detail page, I scraped all of the `.wav` blob url like [this](https://archive.mpi.nl/tla/islandora/object/tla%3A1839_5db33109_854a_47ac_90c9_4e48a6d29643)
3. Now for the hard part, in order to match the `.wav` with the matching `.xml` (caption), I scraped it through [this](https://archive.mpi.nl/tla/islandora/object/tla%3A1839_1cd5563b_cfb4_4b0b_bc05_a18babddc86e?asOfDateTime=2019-03-01T10%3A32%3A30.790Z&f%5B0%5D=cmd.Format%3A%22text%5C/x%5C-eaf%5C%2Bxml%22) blob and find the mosts similar naming for each corresponding file with the `.wav` filename
4. Sadly, the name does not always have the same file naming as the `.wav` file, so I have to recheck one by one. Not to mention there might be `.wav` that does not have any `.xml`

From **4**, I reconsider to elicit my scraping script to the code since it might cause:
1. any trouble if slight UI change happens
2. slow process, pagination scraping of ~200 items, and it does not include the detail page scraping again + caption so the quickest is roughly ~200 * 3 = 600 URLs. Not to mention the time needed might vary more due to user's network speed difference.

Furthermore, I have gone through the discussion in the discord before implementing `alorese` cause this one is a little bit tricky. Got the approval to the such "bulky" method, but for the sake of better result. I can't really open my Discord now, will paste the discussion thread later on.

Comment (ljvmiranda921) [reply seacrowd/sea_datasets/alorese/alorese_url.py:1]: Ah no worries! Thanks for outlining the steps! I guess there's no need to have this be automated, what you did is OK already. Perhaps we should document that process somewhere? Maybe as a docstring in the `alrose_url.py` file? Just a short description to ensure that future folks can replicate where the URLs are coming from. :)

Comment (patrickamadeus) [reply seacrowd/sea_datasets/alorese/alorese_url.py:14]: With pleasure! Please check the newest commit 😄.

**Review by ljvmiranda921 — APPROVED** [review]
LGTM! Let's just wait for @sabilmakbar 's review

File: seacrowd/sea_datasets/alorese/alorese_url.py
Comment (sabilmakbar) [inline seacrowd/sea_datasets/alorese/alorese_url.py:1]: adding timestamp information when you're creating this URL snapshot shd be necessary so others can track the versioning info of URL (in case if something has changed)

File: seacrowd/sea_datasets/alorese/alorese.py
Comment (sabilmakbar) [inline seacrowd/sea_datasets/alorese/alorese.py:36]: consider changing these values:
```suggestion
  journal   = "Oceanic Linguistics",
  publisher = "University of Hawai'i Press",
```

File: seacrowd/sea_datasets/alorese/alorese.py
Comment (sabilmakbar) [inline seacrowd/sea_datasets/alorese/alorese.py:186]: mind reformat this one so it will conform with previous elif statements?

**Review by sabilmakbar — COMMENTED** [review]
init review

File: seacrowd/sea_datasets/alorese/alorese.py
Comment (patrickamadeus) [reply seacrowd/sea_datasets/alorese/alorese.py]: Sure! Please check latest push for the final nitpick 👍

**Comment by sabilmakbar** [comment]
wait I'm going to check it quickly, pardon for late response

File: seacrowd/sea_datasets/alorese/alorese.py
Comment (sabilmakbar) [inline seacrowd/sea_datasets/alorese/alorese.py:98]: may I ask few qns here?

1. overall, the schema we will use in here are T2T for translation of transcription of same audio from Indonesian to Alorese, and the SPText is the ASR version, right?

2. Which language does the audio content has? Indonesian or Alorese? bcs when I looked the code, there's no clear way to tell which lang is available in the transcription (and it's crucial to correctly map the Audio to the correct Transcriptions)

3. For the config name below:
`{_DATASETNAME} _sptext_seacrowd_sptext` and `{_DATASETNAME}_t2t_seacrowd_t2t`, may I know why the naming isn't something like `{_DATASETNAME}_seacrowd_sptext` and `{_DATASETNAME}_seacrowd_t2t`? any justifications here?

4. I saw in the source, the info on `speaker_id` went missing? I thought `source` config should include all columns and informations (and stitch them appropriately if the data scattered into different files and configs, like what you did in your code)

Comment (patrickamadeus) [reply seacrowd/sea_datasets/alorese/alorese.py:98]: Hi @sabilmakbar , thanks for the review and questions!

1. Yes
2. The audio is in alorese, the lang available in transcription is denoted by either `sptext` for alorese or `sptext_trans` for indonesian. The mapping is done in this part
image

3. Because of the subset naming, the testing code (and dataset naming format) that was constructed needs to be in `<DATASET_NAME>_<SUBSET_NAME>_seacrowd_`<SCHEMA>`. It's just the subset name happens to be same as the schema, so it might be confusing. Happy to change if it might be needed.

4. Thanks for the input! Please review the latest commit as I have done the nitpick.

Comment (sabilmakbar) [reply seacrowd/sea_datasets/alorese/alorese.py]: thanks for the prompt reply, @patrickamadeus!

> The audio is in alorese, the lang available in transcription is denoted by either sptext for alorese or sptext_trans for indonesian. The mapping is done in this part

Idk whether we need to create a schema of alorese audio against Indonesian text for the SPText schema. my personal opinion is to remove it since it could be very misleading (ASR schema should provide a text which is an actual transcription as is, not the translated ones)


> Because of the subset naming, the testing code (and dataset naming format) that was constructed needs to be in <DATASET_NAME>_<SUBSET_NAME>_seacrowd_`. It's just the subset name happens to be same as the schema, so it might be confusing. Happy to change if it might be needed.

If I understand this correctly, the subsets for this dataset are only the Alorese version and Indonesian version. The SPText and T2T don't fit properly to the definition of "subset" of a dataset, as only the schema is different.


wdyt? @ljvmiranda921 @holylovenia @SamuelCahyawijaya

Comment (ljvmiranda921) [reply seacrowd/sea_datasets/alorese/alorese.py]: Ah I see, hmm, does it mean we should have a separate data loaders for the SPText and the T2T versions? I don't have a particularly strong opinion on any approach.

Comment (sabilmakbar) [reply seacrowd/sea_datasets/alorese/alorese.py]: > Ah I see, hmm, does it mean we should have a separate data loaders for the SPText and the T2T versions?

No, we can have it in a single dataloader; I think the subset and configuration naming should be modified slightly.

> The audio is in alorese, the lang available in transcription is denoted by either sptext for alorese or sptext_trans for indonesian.

The text information provided in this dataset is sequential; i.e., for every audio file, there is a sequence of annotated texts with their start & end timestamps.

For the Alorese language of text and audio, this can be put in ASR schema (or even a sequential audio split to its annotated text if we want to refine it further).

However, I don't think we should create an ASR schema for the Alorese audio and translated Indonesian annotation since the audio and text are in different languages.

And for T2T of Alorese and Indonesian translation, the existing implementation is correct, just need to reconstruct the configs list.

Therefore, my proposed configs are:
1. Source -- containing Alorese Audio, Alorese Annotation (and its timestamp), and Indonesian Annotation (and its timestamp too)
2. SPText -- containing Alorese Audio & its Annotation (the annotation could be combined into single text per audio -- as previously implemented or recreated using new sequenced schema of text)
3. T2T -- containing Alorese Annotation & translated Indonesian Annotation (we can leave this as full-text translation, not word-to-word or phrase-to-phrase translation)

Comment (patrickamadeus) [reply seacrowd/sea_datasets/alorese/alorese.py]: Hi! Thanks for the suggestions @sabilmakbar !

> Btw, just confirming, do the audio recordings and the transcriptions match word-for-word?

Yes! There are multiple timestamps to indicate when each word is spoken. @holylovenia , to be honest I haven't reviewed substantial sample from the data to determine whether it matched word-for-word or no, but as I listened to the first 10 seconds of 1 particular example, it perfectly matched.

If there is no further suggestion or comment, I will implement the change maximum this weekend, got a bunch of stuff to do first.

Comment (patrickamadeus) [reply seacrowd/sea_datasets/alorese/alorese.py]: Hi @sabilmakbar ! Could you please check on the latest commit? I have done the revision 👍.

> SPText -- containing Alorese Audio & its Annotation (the annotation could be combined into single text per audio -- as previously implemented or recreated using new sequenced schema of text)

For this one, I went on using the previous implementation first.

File: seacrowd/sea_datasets/alorese/alorese.py
Comment (sabilmakbar) [inline seacrowd/sea_datasets/alorese/alorese.py:68]: can we fix this formatting? :D

Comment (patrickamadeus) [reply seacrowd/sea_datasets/alorese/alorese.py]: bueno @sabilmakbar ! it's done, sorry I forgot to delete the old `]` bracket

File: seacrowd/sea_datasets/alorese/alorese.py
Comment (sabilmakbar) [inline seacrowd/sea_datasets/alorese/alorese.py:111]: this `schemas.speech_text_features` is a function, not an object of `datasets.Features`

```suggestion
            features = schemas.speech_text_features()
```

File: seacrowd/sea_datasets/alorese/alorese.py
Comment (sabilmakbar) [inline seacrowd/sea_datasets/alorese/alorese.py:227]: some `value` variable in here isn't a string, found some float that throws error. consider changing it to:

```suggestion
            .agg({"annotation_aol": lambda x: " ".join([str(value) if value is not None else "<NONE>" for value in x]), "annotation_ind": lambda x: " ".join([str(value) if value is not None else "<NONE>" for value in x])})
```

**Comment by sabilmakbar** [comment]
Hi @patrickamadeus, I already put in an updated review. Let both of us know if the suggestion has been addressed, prob both me and LJ need to re-run the whole checking once more to ensure it's already correct since this data loader is quite complex. Thx!

File: seacrowd/sea_datasets/alorese/alorese.py
Comment (patrickamadeus) [reply seacrowd/sea_datasets/alorese/alorese.py]: Thank you for the sharp notice! Looks plausible to me. I'm commiting the suggestion except there is more improvement to be addressed.

**Comment by sabilmakbar** [comment]
Hi @patrickamadeus, all looks good to me. Since LJ said he doesn't have much PC storage left (presumably), I'll proceed with the merge :) (I am able to download all data & subsets and tested it too).

How does it sound, @ljvmiranda921? If that's fine from your end, I'll approve and merge it

**Comment by ljvmiranda921** [comment]
^Yes please feel free to merge! 🙇

**Review by sabilmakbar — APPROVED** [review]
lgtm!

**Pull request merged**
